File: CS/MyComboBoxEdit/MyInvisibleIndexes.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.XtraEditors;



namespace MyComboBoxEdit
{
    class MyInvisibleIndexes
    {
        private ComboBoxEdit comboBox;
        private List<int> contentList;
        private CollectionChangeEventHandler itemsChange;
        private MeasureItemEventHandler itemMeasure;


        public int this[int index]
        {
            get
            {
                return contentList[index];
            }
            set
            {
                contentList[index] = value;
            }
        }



        public MyInvisibleIndexes(ComboBoxEdit comboBox)
        {
            itemsChange = null;
            itemMeasure = null;
            contentList = new List<int>();
            this.comboBox = comboBox;
            Init();
        }



        ~MyInvisibleIndexes()
        {
            if (comboBox != null)
            {
                try
                {
                    comboBox.Properties.MeasureItem -= itemMeasure;
                    comboBox.Properties.Items.CollectionChanged -= itemsChange;
                }
                catch
                {
                }
            }
        }



        private void Init()
        {
            if (comboBox != null)
            {
                if (itemsChange == null)
                {
                    itemsChange = new CollectionChangeEventHandler(OnItemsChanged);
                    comboBox.Properties.Items.CollectionChanged += itemsChange;
                }

                if (itemMeasure == null)
                {
                    itemMeasure = new MeasureItemEventHandler(OnMeasureItem);
                    comboBox.Properties.MeasureItem += itemMeasure;
                }
            }
        }



        private void OnItemsChanged(object sender, CollectionChangeEventArgs e)
        {
            ComboBoxEdit combo = sender as ComboBoxEdit;
            if (combo == null) return;

            int countItems = combo.Properties.Items.Count;
            if (countItems == 0)
            {
                contentList.Clear();
                return;
            }

            int ndx = 0;
            while (ndx < contentList.Count)
            {
                if (contentList[ndx] >= countItems)
                {
                    contentList.RemoveAt(ndx);
                    continue;
                }

                ndx++;
            }
        }



        private void OnMeasureItem(object sender, MeasureItemEventArgs e)
        {
            if (contentList.Contains(e.Index))
            {
                e.ItemHeight = 0;
                e.ItemWidth = 0;
            }
        }



        private void ClearPopup()
        {
            Control control = (comboBox as DevExpress.Utils.Win.IPopupControl).PopupWindow;
            if (control != null)
            {
                control.Parent = null;
                control.Dispose();
            }
        }



        public int Add(int value)
        {
            int iResult = -1;
            if (comboBox == null || contentList.Contains(value)) return iResult;

            try
            {
                ClearPopup();
                contentList.Add(value);
                iResult = contentList.Count - 1;
            }
            catch
            {
            }

            return iResult;
        }



        public void AddRange(int[] indexes)
        {
            if (indexes == null) return;

            for (int i = 0; i < indexes.Length; i++) Add(indexes[i]);
        }



        public void Clear()
        {
            contentList.Clear();
        }



        public bool Contains(int value)
        {
            return contentList.Contains(value);
        }



        public int IndexOf(int value)
        {
            return contentList.IndexOf(value);
        }



        public int Count
        {
            get
            {
                return contentList.Count;
            }
        }



        public void Remove(int value)
        {
            contentList.Remove(value);
        }
    }
}

File: VB/MyComboBoxEdit/MyInvisibleIndexes.vb
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.XtraEditors

Namespace MyComboBoxEdit

    Friend Class MyInvisibleIndexes

        Private comboBox As ComboBoxEdit

        Private contentList As List(Of Integer)

        Private itemsChange As CollectionChangeEventHandler

        Private itemMeasure As MeasureItemEventHandler

        Default Public Property Item(ByVal index As Integer) As Integer
            Get
                Return contentList(index)
            End Get

            Set(ByVal value As Integer)
                contentList(index) = value
            End Set
        End Property

        Public Sub New(ByVal comboBox As ComboBoxEdit)
            itemsChange = Nothing
            itemMeasure = Nothing
            contentList = New List(Of Integer)()
            Me.comboBox = comboBox
            Init()
        End Sub

        Protected Overrides Sub Finalize()
            If comboBox IsNot Nothing Then
                Try
                    RemoveHandler comboBox.Properties.MeasureItem, itemMeasure
                    RemoveHandler comboBox.Properties.Items.CollectionChanged, itemsChange
                Catch
                End Try
            End If
        End Sub

        Private Sub Init()
            If comboBox IsNot Nothing Then
                If itemsChange Is Nothing Then
                    itemsChange = New CollectionChangeEventHandler(AddressOf OnItemsChanged)
                    AddHandler comboBox.Properties.Items.CollectionChanged, itemsChange
                End If

                If itemMeasure Is Nothing Then
                    itemMeasure = New MeasureItemEventHandler(AddressOf OnMeasureItem)
                    AddHandler comboBox.Properties.MeasureItem, itemMeasure
                End If
            End If
        End Sub

        Private Sub OnItemsChanged(ByVal sender As Object, ByVal e As CollectionChangeEventArgs)
            Dim combo As ComboBoxEdit = TryCast(sender, ComboBoxEdit)
            If combo Is Nothing Then Return
            Dim countItems As Integer = combo.Properties.Items.Count
            If countItems = 0 Then
                contentList.Clear()
                Return
            End If

            Dim ndx As Integer = 0
            While ndx < contentList.Count
                If contentList(ndx) >= countItems Then
                    contentList.RemoveAt(ndx)
                    Continue While
                End If

                ndx += 1
            End While
        End Sub

        Private Sub OnMeasureItem(ByVal sender As Object, ByVal e As MeasureItemEventArgs)
            If contentList.Contains(e.Index) Then
                e.ItemHeight = 0
                e.ItemWidth = 0
            End If
        End Sub

        Private Sub ClearPopup()
            Dim control As Control = TryCast(comboBox, DevExpress.Utils.Win.IPopupControl).PopupWindow
            If control IsNot Nothing Then
                control.Parent = Nothing
                control.Dispose()
            End If
        End Sub

        Public Function Add(ByVal value As Integer) As Integer
            Dim iResult As Integer = -1
            If comboBox Is Nothing OrElse contentList.Contains(value) Then Return iResult
            Try
                ClearPopup()
                contentList.Add(value)
                iResult = contentList.Count - 1
            Catch
            End Try

            Return iResult
        End Function

        Public Sub AddRange(ByVal indexes As Integer())
            If indexes Is Nothing Then Return
            For i As Integer = 0 To indexes.Length - 1
                Add(indexes(i))
            Next
        End Sub

        Public Sub Clear()
            contentList.Clear()
        End Sub

        Public Function Contains(ByVal value As Integer) As Boolean
            Return contentList.Contains(value)
        End Function

        Public Function IndexOf(ByVal value As Integer) As Integer
            Return contentList.IndexOf(value)
        End Function

        Public ReadOnly Property Count As Integer
            Get
                Return contentList.Count
            End Get
        End Property

        Public Sub Remove(ByVal value As Integer)
            contentList.Remove(value)
        End Sub
    End Class
End Namespace
